### DISCUSSION TITLE
How the new cascade mechanism works

### DISCUSSION THREAD

**Discussion by pmario** [TiddlyWiki/TiddlyWiki5#6280] [General]
## Info

The new version v5.2.1 will have a new Info tab in the [$:/ControlPanel](https://tiddlywiki.com/prerelease/#%24%3A%2FControlPanel)

The main thing to know about the new `[cascade[]]` filter operator is, that "**it tests 1 tiddler after the other and stops if there is a result other than an emptyString**"

The TW story river in the doc edition has 3 filters. 

 - 2 default filters that come with the core
   - If a tiddler is tagged: `$:/tags/StoryTiddlerTemplateFilter` it will be used to define the "Story Cascade"
   - This makes it easy for developers and users to add new filters to the story cascade.
 - 1 that comes with the edition and is a tiddler tagged: `$:/tags/StoryTiddlerTemplateFilter` **and** 
   - has a list-before field to define the order in the cascade
   - The "Custom Story Tiddler Template Demo Filter" contains a filter condition
   - This condition returns the name of the template-tiddler that should be used if a tiddler is tagged: $:/tags/TiddlerList
   - How the condition looks like is completely up the user. 

## Story River Definition

![grafik](https://user-images.githubusercontent.com/374655/143400074-cfea491a-34eb-43a1-bd79-46ea4538cd7b.png)

**Comment by pmario** [comment]
## Tiddler Color Attibute

![grafik](https://user-images.githubusercontent.com/374655/143409356-32056d31-f259-4dd6-baee-cefbade5e6d8.png)

**Reply by pmario** [reply]
The filter in step 2 has been updated to: `[[$:/config/DefaultTiddlerColour]has[text]get[text]trim[]]` which should now return a colour name or colour value stored in the `$:/config/DefaultTiddlerColour` tiddler

**Reply by AnthonyMuscio** [reply]
Mario can that not be simplified to the following?

`[[$:/config/DefaultTiddlerColour]get[text]trim[]]`

No need to change, I just found this works in some cases.

**Comment by pmario** [comment]
## Tiddler Icon

![grafik](https://user-images.githubusercontent.com/374655/143412879-36d5b174-290f-4ea3-ad33-829b48b7ffe4.png)

**Comment by pmario** [comment]
## View Template Body

add 1)
There are a bunch of system-tiddlers that will be shown as code instead of being rendered, with some exceptions for $:/core/readme .. and some others. Since only 1 template output is needed the last filter uses a limit[1]: `+[limit[1]then[$:/core/ui/ViewTemplate/body/code]]`

add 2)
This filter shows the template that we know from the $:/Import tiddler. So every tiddler that has a `plugin-type: import` will be treated the same way. This may be interesting for plugin authors. 

add 3) 
This filter checks if the the `plugin-type` field exists and it will show the template we know from eg: [$:/core](https://tiddlywiki.com/prerelease/#%24%3A%2Fcore)

add 4)
If the field `hide-body` exists, it will show a _new_ "blank" template --- It hides the body

add 5) 
Since this filter always returns the tiddler title "$:/core/ui/ViewTemplate/body/default" it's the default 

![grafik](https://user-images.githubusercontent.com/374655/143413390-014342a7-0440-45db-96f3-e2883658c82b.png)

**Comment by pmario** [comment]
## View Template Title

![grafik](https://user-images.githubusercontent.com/374655/143434250-9877e94b-9169-465c-afa8-357aa88e2bc3.png)

**Comment by pmario** [comment]
## More Info

See some more examples at: https://talk.tiddlywiki.org/t/lets-have-some-fun-with-tiddlywiki-5-2-0-default-tiddler-icon/1595/2?u=pmario

## Core tiddlers that use the `:cascade` filter atm

Search for `:cascade` in **AdvancedSearch : Shadows** tab

 - $:/core/macros/tag
 - $:/core/ui/Components/tag-link
 - $:/core/ui/EditTemplate/body
 - $:/core/ui/EditTemplate/tags
 - $:/core/ui/StoryTiddlerTemplate
 - $:/core/ui/TagPickerTagTemplate
 - $:/core/ui/TagTemplate
 - $:/core/ui/TiddlerIcon
 - $:/core/ui/ViewTemplate/body
 - $:/core/ui/ViewTemplate/title

**Comment by Jermolene** [comment]
>  If you create a tiddler tagged: $:/tags/TiddlerList you can only delete the tiddler with the $:/AdvancedSearch : Filter tab, since the template has now toolbar buttons. ... :/ ... We will need to change that ...

Thanks @pmario I've made a fix in be3018fe3b098ce2783d4f7f3e373990b79424c1

**Comment by AnthonyMuscio** [comment]
Thanks Mario, very clear.

So as  to help my writing this I am using the word "element" (for want of a better term) to describe tiddlers and content which are used to display a tiddler eg title, subtitle, body etc... and others that may be added through the tags $:/tags/ViewTemplate or $:/tags/EditTemplate and now these new mechanisiums.

Can you tell me if I am right in assuming?
- We currently have no cascade configured for the Body Edit? (Although this effectively occurs with different type fields)
  - I can see we can leverage the story tiddler cascade to introduce a different edit template conditionally.
  - but see note 1 below
-  In the current solution the cascades ONLY address the storyTiddler template, colour, icon, view Body, view title.
   - So at present there is no configurable way to address the edit templates or other "view or edit template items" as tagged $:/tags/ViewTemplate or $:/tags/EditTemplate and the toolbars within them?
   - Note These tags are not only used by the core but plugins, macros etc... who would add something the view or edit templates.
   - but see note 1 below

1. To achieve this we can now use the "story tiddler cascade" and introduce alternative view and edit templates to address these other elements!

Observation;

Presently in in $:/core/ui/PageTemplate/story the View and Edit Templates are configured in $:/config/ui/ViewTemplate  and $:/config/ui/EditTemplate and I can see this setting is still honoured. Because this is another way to alter the effective View and Edit templates it may be valuable to annotate this fact in the new cascade tabs, specifically the "story tiddler" cascade tab.

In closing
Mario, You have being very helpful here, as usual.

To explain: I have taken a close interest in this cascade feature, and core UI mods because I am keen to get a little more granular control over the view and edit templates, and toolbars therein, to better handle different "tiddler types" as well as having local and global tiddler modes for the different tiddler types (eg a tiddler in read only mode will hide its ability to be edited etc.. perhaps with a different toolbar). 

Granular control over these elements has always being possible, the desire is to achieve it without modifying core tiddlers for portability and compatibility with other solutions and plugins. Hence the desire to improve the hack-ability of the core UI. It looks like we are 80% of the way, and 100% if we are happy to build alternate view and edit templates and use $:/tags/StoryTiddlerTemplateFilter

Thanks again Mario.

**Reply by pmario** [reply]
>We currently have no cascade configured for the Body Edit? (Although this effectively occurs with different type fields)
>
> - I can see we can leverage the story tiddler cascade to introduce a different edit template conditionally.

Those 2 tiddlers use the :cascase filter at the moment. 

 - $:/core/ui/EditTemplate/body
 - $:/core/ui/EditTemplate/tags

We are still exploring the possibilities ... I'll try to recreate the mechanism described at: https://talk.tiddlywiki.org/t/multilanguage-tiddlywiki-any-existing-resources/1454/26?u=pmario with the new cascade mechanism, to see what's missing and how can we improve the EditTemplate "hackability"

**Reply by pmario** [reply]
>... tagged $:/tags/ViewTemplate or $:/tags/EditTemplate and the toolbars within them?

The toolbars have their own visibility mechanism atm. ... Settings can be changed at the **ControlPanel : Appearence : Toolbars**. 

So plugin authors can add or remove visibility with the according `$:/config/EditToolbarButtons/Visibility/xxx` tiddlers

The `$:/config/TextEditor/EnableToolbar` is currently used to enable / disable to toolbar as a whole. ... IMO this part of the code may be a candidate for a new cascade. ... 

Using cascade filters usually also make the core templates code simpler to read .... So there is a lot of room for improvement.

**Reply by pmario** [reply]
>Presently in in $:/core/ui/PageTemplate/story the View and Edit Templates are configured in $:/config/ui/ViewTemplate and $:/config/ui/EditTemplate and I can see this setting is still honoured. Because this is another way to alter the effective View and Edit templates it may be valuable to annotate this fact in the new cascade tabs, specifically the "story tiddler" cascade tab.

Those panels are dynamically created. 

The code is at: 

https://github.com/Jermolene/TiddlyWiki5/blob/c11493ec37dc587b1cdf7abe78a73c51bdb5136f/core/ui/StoryTiddlerTemplate.tid#L1-L3

It may be possible to have the default setting in the cascade too, so it will show up in the ControlPanel. ... This may be a new PR.

**Reply by pmario** [reply]
>To explain: I have taken a close interest in this cascade feature, and core UI mods because I am keen to get a little more granular control over the view and edit templates, and toolbars therein, to better handle different "tiddler types" as well as having local and global tiddler modes for the different tiddler types (eg a tiddler in read only mode will hide its ability to be edited etc.. perhaps with a different toolbar).

The `:cascade` filter system has been built for exactly that problem. ... As we see atm, it can help to make templates easier to modify and also have a cleaner code. ... If it really holds, we can only see in the future and where the edge cases are for plugin authors. 

>Granular control over these elements has always being possible, the desire is to achieve it without modifying core tiddlers for portability and compatibility with other solutions and plugins. Hence the desire to improve the hack-ability of the core UI. It looks like we are 80% of the way, and 100% if we are happy to build alternate view and edit templates and use $:/tags/StoryTiddlerTemplateFilter

As I wrote, I'll try to recreate the "multinie field editor" functionality without the need to modify the core tiddlers. ... I'm sure we will need to add some more cascades in the edit template to achieve this flexibility. 

-----

At the moment power users modify the ViewTemplate to edit tiddler content. .. I think that's the wrong way. ... 

IMO the end-result should be, that there is no difference between view- and edit-mode. I think if we use cascade filters in a clever way we can make that happen.

**Reply by AnthonyMuscio** [reply]
In relation to your side by side editor, I realised the other day that we can actually design new "editor previews" and need to as I will explain, this is helped by the new tv-tiddler-preview variable to make preview "sticky for each tiddler" when editing.

Here is some things I have learned that may prove useful to your language work and my template vision.

Whilst considering customising the view and edit templates I realised that edit preview only applies to the body, that is we see a preview of the text field only and the default body template, this is certainly going to be more of an issue with the extended template hack-ability. I found a similar problem in the past, with the open in new window, also only rendering the text field, not applying other critical view template elements. As a result I made the attached button that allow Open in new window with the full viewtemplate applied, and even edit in new window.

[new-windows.json.txt](https://github.com/Jermolene/TiddlyWiki5/files/7608787/new-windows.json.txt) pity we can drop JSON files here.

So I think I may build a new preview that is aware of the View (and perhaps Edit) Templating.

**Why is this relevant to you?**
Once a user has used your dual edit to create two or more different languages perhaps they could select the different "language" translations in the edit preview. You could perhaps build these alt-language pre-viewers.It may make it easier for someone to read the alternative language while composing the new language tiddler.

**Other possibilities with your dual edit pain**
It would be nice to edit a system notes tiddler about the current tiddler in the second edit panel, and or preview such system notes for any given tiddler in the preview mechanism. So It would be nice to get these co-operating, perhaps even extend it so others can add additional pains when in edit mode, for example one may have a form tiddler you are writing, in the right hand side you could see a matching and interactive instance of the form on the right, a tiddler with that form applied to it, or the other way around, or just the form applied to a selected tiddler.

If you need clarification please ask.

**Reply by AnthonyMuscio** [reply]
Mario, 
On the toolbar configuration, I have happily built additional toolbars, and whilst I do make use of visibility and conditional buttons there are a few cases where being able to swap whole toolbars is helpful. A feature that did not come across from TWC.

Using a cascade on individual buttons would be nice, but I still think there is a place for alternative toolbars and the application via the template mechanism. This again may be even more desirable if people build fundamentally different templates, such as to drive a read only view etc...

I dream of making a "new toolbar" plugin that can be removed after building a new toolbar.

**Comment by saqimtiaz** [comment]
A few points that might be helpful:

* We can make further changes to the templates and introduce more cascade runs where needed after the 5.2.1 release as well. That is, the templates are not locked after the release. None of the changes introduced now constrain us from adding more flexibility to the templates by making further use of cascades.
* There have been discussions around using cascades to determine the editor used for the tiddler text field. Currently this is all handled in the `$edit` widget which decides which editor to use for each tiddler according to the `$:/config/EditorTypeMappings/` config tiddlers. 
  * Breaking this out into wikitext has been desirable even independently of the introduction of the `:cascade` mechanism. 
  * However, this will need to be approached separately and carefully as there are lot of aspects to consider including backwards compatibility. Amongst other things, making changes in this area will break plugins such as EditCompText.
